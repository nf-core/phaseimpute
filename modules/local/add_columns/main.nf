process ADD_COLUMNS {
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/gawk:5.3.0' :
        'biocontainers/gawk:5.3.0' }"

    input:
    tuple val(meta), path(input)

    output:
    tuple val(meta), path('*.txt'), emit: txt
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Find the header line
    HEADER_STR="#Genotype concordance by allele frequency bin (Variants: SNPs + indels)"
    HEADER_LINE=\$(grep -n -m 1 "^\${HEADER_STR}" $input | cut -d: -f1 )
    HEADER_START=\$((HEADER_LINE + 1))

    tail -n +\$HEADER_START $input | \\
    awk 'NR==1{\$(NF+1)="ID"} NR>1{\$(NF+1)="${meta.id}"}1' | \\
    awk 'NR==1{\$(NF+1)="Depth"} NR>1{\$(NF+1)="${meta.depth}"}1' | \\
    awk 'NR==1{\$(NF+1)="GPArray"} NR>1{\$(NF+1)="${meta.gparray}"}1' | \\
    awk 'NR==1{\$(NF+1)="Tools"} NR>1{\$(NF+1)="${meta.tools}"}1' | \\
    awk 'NR==1{\$(NF+1)="Panel"} NR>1{\$(NF+1)="${meta.panel}"}1' > \\
    ${prefix}.txt


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gawk: \$(awk -Wversion | sed '1!d; s/.*Awk //; s/,.*//')
    END_VERSIONS
    """
}
