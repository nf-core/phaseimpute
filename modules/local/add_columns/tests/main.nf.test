nextflow_process {

    name "Test Process ADD_COLUMNS"
    script "../main.nf"
    process "ADD_COLUMNS"

    tag "modules"
    tag "modules_local"
    tag "add_columns"
    tag "gawk"

    test("Add columns to txt file") {

        when {
            process {
                """
                file = Channel.of("# Other content\\ntest\\ntest1\\n#Genotype concordance by allele frequency bin (Variants: SNPs + indels)\\nCol1\\tCol2\\nA\\t1\\nB\\t2")
                    .collectFile(name: 'sample.txt', newLine: true)
                input[0] = Channel.of([[id:"MyId", depth:2, gparray:"SNPArray", tools:"Glimpse", panel:"1000GP"]]).combine(file)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert snapshot(path(process.out.txt[0][1]).readLines()).match("Content") }
            )
        }
    }
    test("Add columns to txt file with missing fields") {

        when {
            process {
                """
                file = Channel.of("# Other content\\ntest\\ntest1\\n" +
                    "#Genotype concordance by allele frequency bin (Variants: SNPs + indels)\\n" +
                    "Col1\\tCol2\\nA\\t1\\nB\\t2")
                    .collectFile(name: 'sample.txt', newLine: true)
                input[0] = Channel.of([[id:"MyId", tools:"Glimpse"]]).combine(file)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
                { assert snapshot(path(process.out.txt[0][1]).readLines()).match("ContentWithNA") }
            )
        }
    }
}
