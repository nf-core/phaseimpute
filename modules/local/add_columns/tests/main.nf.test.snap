{
    "Content": {
        "content": [
            [
                "Col1 Col2 ID Depth GPArray Tools Panel",
                "A 1 MyId 2 SNPArray Glimpse 1000GP",
                "B 2 MyId 2 SNPArray Glimpse 1000GP"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T14:43:39.482844695"
    },
    "ContentWithNA": {
        "content": [
            [
                "Col1 Col2 ID Depth GPArray Tools Panel",
                "A 1 MyId null null Glimpse null",
                "B 2 MyId null null Glimpse null"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T14:44:39.395698015"
    },
    "Add columns to txt file": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "MyId",
                            "depth": 2,
                            "gparray": "SNPArray",
                            "tools": "Glimpse",
                            "panel": "1000GP"
                        },
                        "MyId.txt:md5,da3aa04353303c5566d1bea2dbc8f883"
                    ]
                ],
                "1": [
                    "versions.yml:md5,b25787d2ba80184a94de82cf5cfe8e7a"
                ],
                "txt": [
                    [
                        {
                            "id": "MyId",
                            "depth": 2,
                            "gparray": "SNPArray",
                            "tools": "Glimpse",
                            "panel": "1000GP"
                        },
                        "MyId.txt:md5,da3aa04353303c5566d1bea2dbc8f883"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b25787d2ba80184a94de82cf5cfe8e7a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T14:43:39.390342082"
    },
    "Add columns to txt file with missing fields": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "MyId",
                            "tools": "Glimpse"
                        },
                        "MyId.txt:md5,87ae2dd6e6451eee25c04e23f7219fa5"
                    ]
                ],
                "1": [
                    "versions.yml:md5,b25787d2ba80184a94de82cf5cfe8e7a"
                ],
                "txt": [
                    [
                        {
                            "id": "MyId",
                            "tools": "Glimpse"
                        },
                        "MyId.txt:md5,87ae2dd6e6451eee25c04e23f7219fa5"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b25787d2ba80184a94de82cf5cfe8e7a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T14:44:39.317216249"
    }
}